test/built-ins/Atomics/wait/bigint/false-for-timeout-agent.js

```diff
@@ -31,21 +33,31 @@ $262.agent.start(`
 
   $262.agent.receiveBroadcast(function(sab) {
     const i64a = new BigInt64Array(sab);
+    Atomics.add(i64a, ${RUNNING}, 1n);
+
     const before = $262.agent.monotonicNow();
-    $262.agent.report(Atomics.wait(i64a, 0, 0n, false));
-    $262.agent.report(Atomics.wait(i64a, 0, 0n, valueOf));
-    $262.agent.report(Atomics.wait(i64a, 0, 0n, toPrimitive));
-    $262.agent.report($262.agent.monotonicNow() - before);
+    const status1 = Atomics.wait(i64a, 0, 0n, false);
+    const status2 = Atomics.wait(i64a, 0, 0n, valueOf);
+    const status3 = Atomics.wait(i64a, 0, 0n, toPrimitive);
+    const duration = $262.agent.monotonicNow() - before;
+
+    $262.agent.report(status1);
+    $262.agent.report(status2);
+    $262.agent.report(status3);
+    $262.agent.report(duration);
     $262.agent.leaving();
   });
 `);
 
 const i64a = new BigInt64Array(
-  new SharedArrayBuffer(BigInt64Array.BYTES_PER_ELEMENT * 8)
+  new SharedArrayBuffer(BigInt64Array.BYTES_PER_ELEMENT * 4)
 );
 
 $262.agent.broadcast(i64a.buffer);
-$262.agent.sleep(100);
+$262.agent.waitUntil(i64a, RUNNING, 1n);
+
+// Try to yield control to ensure the agent actually started to wait.
+$262.agent.tryYield();
```

Comment:
@anba I had worked to eliminate as much reliance on "sleep" as possible, by making codifying `waitUntil` and implementing it wherever it was necessary to wait for all agents to "report" that they are moving into a waiting state. `tryYield` is just reintroducing the "sleep" _in addition_ to the "wait until" mechanism. Can you help me understand why that extra step is necessary? If I had to guess, I _think_ that you're sleeping the main thread in order to allow the agents' execution to reach the end of calling `Atomics.wait(...)`? If that's true, then `262.agent.tryYield();` serves as essentially 1/2 of what `waitUntil` is meant to provide and I think that the implementation of `$262.agent.waitUntil` should itself call `$262.agent.tryYield();`, instead of putting the burden of remembering to _also_ call `$262.agent.tryYield();` on test authors (and future maintainers). 

If you agree with me, I will do the work of updating the test material. WDYT?

Reply:
Yes, `tryYield` is necessary for the case when the agent thread was interrupted between its call to `Atomics.add(i32a, ${RUNNING}, 1);` and the call to `Atomics.wait(i32a, 0, ...)`. This case happens all the time when the system is under heavy load, so not performing `tryYield` is sadly not an option. (We also want to have `waitUntil`, because that allows us to reduce the sleep time in `tryYield`.) 

But if a specific test262 host has better a mean to implement `tryYield` than just sleeping for some time, I'd currently recommend that this host provides their own `tryYield` implementation. (I haven't yet investigated what is necessary to enable the Atomics tests in a browser environment, but the patch in https://bugzilla.mozilla.org/show_bug.cgi?id=1349863 gives the impression that quite a few changes for `$test262.agent` are necessary. That's why I don't yet want to make `tryYield` a part of the requirement host binding for `$test262.agent`, because browsers may need a complete different approach than `tryYield`.)

Most, but not all tests, call `tryYield` directly after `waitUntil`, therefore I haven't moved `tryYield` into `waitUntil`. For example "waiterlist-block-indexedposition-wake.js" (`waitUntil` not followed by `tryYield`) or "waiterlist-order-of-operations-is-fifo.js" (`tryYield` not directly preceded by `waitUntil`).

And I also agree that reducing direct calls to `$test262.agent.sleep` in test cases is a useful goal. In fact my PR actually removed all direct calls to `$test262.agent.sleep`, so the tests are now only calling it indirectly. :smile: